### PR TITLE
[WB-6947] Flush log data without finish with service

### PR DESCRIPTION
**Pull request by kptkin**
Fixes [WB-6947](https://wandb.atlassian.net/browse/WB-6947)

Description
-----------
What does the PR do?

This PR moves the partial history logic to the internal process. This PR allows for runs that use the `wandb-service` path to flush their remaining history without explicitly calling `wandb.finish`.

This is the first PR in series of PR that move the step tracking logic from the user process to the internal process.

Testing
-------
How was this PR tested?
This PR should keep functionality, it is tested using the existing tests.  

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/wandb_torch.py [hunk 1]
```diff
@@ -96,6 +95,7 @@ def add_log_hooks_to_pytorch_module(
         if name is not None:
             prefix = prefix + name
 
+        # TODO should we remove this?
         if jupyter_run:
```
Comment (kptkin) [inline wandb/wandb_torch.py:99]: should we remove this?

File: wandb/integration/tensorboard/log.py [hunk 2]
```diff
@@ -300,6 +301,14 @@ def log(tf_summary_str_or_pb, history=None, step=0, namespace="", **kwargs):
...
+            if history is None:
+                wandb.log({})
+            else:
+                history.add({}, **kwargs)
+
         STEPS["global"]["last_log"] = timestamp
-    history._row_update(log_dict)
+
+    if history is None:
+        wandb.log(log_dict, commit=False)
```
Comment (raubitsj) [inline wandb/integration/tensorboard/log.py:312]: Lets make an internal version of run._log() so we can know that this came from a special integration.

File: wandb/sdk/internal/handler.py [hunk 3]
```diff
@@ -454,16 +484,46 @@ def handle_history(self, record: Record) -> None:
...
+                    record.request.partial_history, self._partial_history
+                )
+
+            for k, v in self._partial_history.items():
+                item = record.request.partial_history.item.add()
+                item.key = k
+                item.value_json = json.dumps(v)
+
+            self._history_update(record.request.partial_history, self._partial_history)
+            self._dispatch_record(record)
```
Comment (raubitsj) [inline wandb/sdk/internal/handler.py:510]: I think you should dispatch a HistoryRecord and leave sender untouched with respect to partial history?

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -320,8 +314,8 @@ def __init__(
             self._summary_get_current_summary_callback,
         )
         self.summary._set_update_callback(self._summary_update_callback)
-        self.history = wandb_history.History(self)
-        self.history._set_callback(self._history_callback)
+        self.history_step = 0
+        self._torch_history: Union["wandb.wandb_torch.TorchHistory", None] = None
```
Comment (raubitsj) [inline wandb/sdk/wandb_run.py:318]: Optional

Comment (kptkin) [reply wandb/sdk/wandb_run.py:318]: I think i just copied it... good catch

File: wandb/sdk/internal/handler.py (hunk 3 above)
Comment (kptkin) [reply wandb/sdk/internal/handler.py:510]: Yeah... I don't remember whey I did it this way... but in theory it seems like the cleaner way to do it

File: wandb/integration/tensorboard/log.py (hunk 2 above)
Comment (kptkin) [reply wandb/integration/tensorboard/log.py:312]: good idea!

File: wandb/wandb_torch.py [hunk 5]
```diff
@@ -96,9 +92,6 @@ def add_log_hooks_to_pytorch_module(
         if name is not None:
             prefix = prefix + name
 
-        if jupyter_run:
```
Comment (raubitsj) [inline wandb/wandb_torch.py:99]: Im just curious. did you look into why this was here and what it was accomplishing before?

**Review by raubitsj — APPROVED** [review]
PERFECT!

File: wandb/wandb_torch.py (hunk 5 above)
Comment (kptkin) [reply wandb/wandb_torch.py:99]: I couldn't find what is was doing... I saw it is used to recover history... 
I guess I can keep it and just use it as a run...

File: wandb/sdk/wandb_watch.py [hunk 6]
```diff
@@ -86,19 +88,16 @@ def watch(
...
             prefix = "graph_%i" % global_idx
 
-        wandb.run.history.torch.add_log_hooks_to_pytorch_module(
+        wandb.run._torch.add_log_hooks_to_pytorch_module(
             model,
             log_parameters=log_parameters,
             log_gradients=log_gradients,
             prefix=prefix,
             log_freq=log_freq,
-            jupyter_run=wandb.run if in_jupyter else None,
```
Comment (kptkin) [inline wandb/sdk/wandb_watch.py:95]: it is using the same `wandb.run` so that's why i though it's ok to remove

**Pull request merged**